service/updates/downloader.go

```diff
@@ -0,0 +1,344 @@
+func (d *Downloader) copyMatchingFilesFromCurrent(currentFiles map[string]File) error {
+	// Make sure new dir exists
+	_ = os.MkdirAll(d.dir, defaultDirMode)
+
+	for _, a := range d.bundle.Artifacts {
+		currentFile, ok := currentFiles[a.Filename]
+		if ok && currentFile.Sha256() == a.SHA256 {
+			// Read the content of the current file.
+			content, err := os.ReadFile(currentFile.Path())
+			if err != nil {
+				return fmt.Errorf("failed to read file %s: %w", currentFile.Path(), err)
+			}
+
+			// Check if the content matches the artifact hash
+			expectedHash, err := hex.DecodeString(a.SHA256)
+			if err != nil || len(expectedHash) != sha256.Size {
+				return fmt.Errorf("invalid artifact hash %s: %w", a.SHA256, err)
+			}
+			hash := sha256.Sum256(content)
+			if !bytes.Equal(expectedHash, hash[:]) {
+				return fmt.Errorf("expected and file hash mismatch: %s", currentFile.Path())
+			}
+
+			// Create new file
+			destFilePath := filepath.Join(d.dir, a.Filename)
+			err = os.WriteFile(destFilePath, content, a.GetFileMode())
+			if err != nil {
+				return fmt.Errorf("failed to write to file %s: %w", destFilePath, err)
+			}
+			log.Debugf("updates: file copied from current version: %s", a.Filename)
+		}
+	}
+	return nil
+}
+
+func (d *Downloader) downloadAndVerify(ctx context.Context) error {
+	// Make sure we have the bundle file parsed.
+	err := d.parseBundle()
+	if err != nil {
+		log.Errorf("updates: invalid update bundle file: %s", err)
+	}
+
+	// Make sure dir exists
+	_ = os.MkdirAll(d.dir, defaultDirMode)
+
+	for _, artifact := range d.bundle.Artifacts {
+		filePath := filepath.Join(d.dir, artifact.Filename)
+
+		// Check file is already downloaded and valid.
+		exists, _ := checkIfFileIsValid(filePath, artifact)
+		if exists {
+			log.Debugf("updates: file already downloaded: %s", filePath)
+			continue
+		}
+
+		// Download artifact
+		err := d.processArtifact(ctx, artifact, filePath)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *Downloader) processArtifact(ctx context.Context, artifact Artifact, filePath string) error {
+	providedHash, err := hex.DecodeString(artifact.SHA256)
+	if err != nil || len(providedHash) != sha256.Size {
+		return fmt.Errorf("invalid provided hash %s: %w", artifact.SHA256, err)
+	}
+
+	// Download
+	log.Debugf("updates: downloading file: %s", artifact.Filename)
+	content, err := d.downloadFile(ctx, artifact.URLs)
+	if err != nil {
+		return fmt.Errorf("failed to download artifact: %w", err)
+	}
+
+	// Decompress
+	if artifact.Unpack != "" {
+		content, err = unpack(artifact.Unpack, content)
+		if err != nil {
+			return fmt.Errorf("failed to decompress artifact: %w", err)
+		}
+	}
+
+	// Verify
+	hash := sha256.Sum256(content)
+	if !bytes.Equal(providedHash, hash[:]) {
+		return fmt.Errorf("failed to verify artifact: %s", artifact.Filename)
+	}
```

Comment:
Move decompression and verification to `downloadFile` for more resilience. This would allow us to ignore URLs that ship the wrong file (or a "200 OK" error page, or redirect).